core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java

```diff
@@ -0,0 +1,322 @@
+   */
+  private UnderFileSystemBlockReader(UnderFileSystemBlockMeta blockMeta, boolean noCache,
+      BlockStore alluxioBlockStore) {
+    mFileBufferSize = Configuration.getBytes(PropertyKey.WORKER_FILE_BUFFER_SIZE);
+    mBlockMeta = blockMeta;
+    mAlluxioBlockStore = alluxioBlockStore;
+    mNoCache = noCache;
+    mInStreamPos = -1;
+    mBlockMeta.setBlockReader(this);
+  }
+
+  /**
+   * Initializes the reader. This is only called in the factory method.
+   *
+   * @param offset the position within the block to start the read
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the UFS block store
+   * @throws IOException if an I/O related error occur
+   */
+  private void init(long offset) throws BlockDoesNotExistException, IOException {
+    UnderFileSystem ufs = UnderFileSystem.Factory.get(mBlockMeta.getUnderFileSystemPath());
+    ufs.connectFromWorker(
+        NetworkAddressUtils.getConnectHost(NetworkAddressUtils.ServiceType.WORKER_RPC));
+    if (!ufs.isFile(mBlockMeta.getUnderFileSystemPath())) {
+      throw new BlockDoesNotExistException(
+          ExceptionMessage.UFS_PATH_DOES_NOT_EXIST.getMessage(mBlockMeta.getUnderFileSystemPath()));
+    }
+
+    updateUnderFileSystemInputStream(offset);
+    updateBlockWriter(offset);
+  }
+
+  @Override
+  public ReadableByteChannel getChannel() {
+    throw new UnsupportedOperationException("UFSFileBlockReader#getChannel is not supported");
+  }
+
+  @Override
+  public long getLength() {
+    return mBlockMeta.getBlockSize();
+  }
+
+  @Override
+  public ByteBuffer read(long offset, long length) throws IOException {
+    Preconditions.checkState(!mClosed);
+    updateBlockWriter(offset);
+    updateUnderFileSystemInputStream(offset);
+
+    long bytesToRead = Math.min(length, mBlockMeta.getBlockSize() - offset);
+    if (bytesToRead <= 0) {
+      return null;
+    }
+    byte[] data = new byte[(int) bytesToRead];
+    int bytesRead = 0;
+    if (mUnderFileSystemInputStream != null) {
+      while (bytesRead < bytesToRead) {
+        int read =
+            mUnderFileSystemInputStream.read(data, bytesRead, (int) (bytesToRead - bytesRead));
+        if (read == -1) {
+          break;
+        }
+        bytesRead += read;
+      }
+    }
+    mInStreamPos += bytesRead;
+
+    // We should always read the number of bytes as expected since the UFS file length (hence block
+    // size) should be always accurate.
+    Preconditions
+        .checkState(bytesRead == bytesToRead, PreconditionMessage.NOT_ENOUGH_BYTES_READ.toString(),
+            bytesRead, bytesToRead, mBlockMeta.getUnderFileSystemPath());
+    if (mBlockWriter != null && mBlockWriter.getPosition() < mInStreamPos) {
+      Preconditions.checkState(mBlockWriter.getPosition() >= offset);
+      ByteBuffer buffer = ByteBuffer.wrap(data, (int) (mBlockWriter.getPosition() - offset),
+          (int) (mInStreamPos - mBlockWriter.getPosition()));
+      mBlockWriter.append(buffer.duplicate());
```

Comment:
I understand. My question is since no one is using `buffer`, why do we need to make the duplicated copy of state, why not using `append(buffer)` directly?